# FINAL PROJECT

Người thực hiện

    Họ tên: Nguyễn Đình Tuyên. MSSV: 1712888

    Họ tên: Đặng Đỗ Huỳnh Như. MSSV: 18120219

## Import các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 1. Thu thập dữ liệu

- Dữ liệu trong file "data_scientist_salary.csv" được lấy trên kaggle mô tả về mức lương của nhà khoa học dữ liệu tại các bang của nước Mỹ. 

- License của dữ liệu là CC0: Public domain. Do đó, dữ liệu cho phép sử dụng.

- Dữ liệu được thu thập từ trang web GlassDoor bằng cách sử dụng trình quét Selenium.

## 2. Khám phá dữ liệu

### Đọc dữ liệu từ file csv vào dataframe

In [2]:
ds_salary_df = pd.read_csv('data_scientist_salary.csv')
ds_salary_df.head()

index                  Job Title              Salary Estimate  \
0      0             Data Scientist   $53K-$91K (Glassdoor est.)   
1      1  Healthcare Data Scientist  $63K-$112K (Glassdoor est.)   
2      2             Data Scientist   $80K-$90K (Glassdoor est.)   
3      3             Data Scientist   $56K-$97K (Glassdoor est.)   
4      4             Data Scientist  $86K-$143K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Data Scientist\nLocation: Albuquerque, NM\nEdu...     3.8   
1  What You Will Do:\n\nI. General Summary\n\nThe...     3.4   
2  KnowBe4, Inc. is a high growth information sec...     4.8   
3  *Organization and Job ID**\nJob ID: 310709\n\n...     3.8   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

                                 Company Name         Location  \
0                      Tecolote Research\n3.8  Albuquerque, NM   
1  University of Maryland Medical System\n3.4    Linthicum, MD   
2                                KnowBe4\n4.8   Clearwater, FL   
3                                   PNNL\n3.8     Richland, WA   
4                     Affinity Solutions\n2.9     New York, NY   

     Headquarters          Size  Founded  ... tensor hadoop tableau bi flink  \
0      Goleta, CA   501 - 1000      1973  ...      0      0       1  1     0   
1   Baltimore, MD       10000+      1984  ...      0      0       0  0     0   
2  Clearwater, FL   501 - 1000      2010  ...      0      0       0  0     0   
3    Richland, WA  1001 - 5000      1965  ...      0      0       0  0     0   
4    New York, NY     51 - 200      1998  ...      0      0       0  0     0   

   mongo  google_an   job_title_sim  seniority_by_title  Degree  
0      0          0  data scientist                  na       M  
1      0          0  data scientist                  na       M  
2      0          0  data scientist                  na       M  
3      0          0  data scientist                  na      na  
4      0          0  data scientist                  na      na  

[5 rows x 42 columns]

- Bỏ cột **index**, sử dụng theo index của pandas, vì theo quan sát ta thấy cột này chỉ chỉ mục của dòng nhưng lại không liên tục.

In [3]:
ds_salary_df = ds_salary_df.drop(columns = ['index'])

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

- Biến num_rows và num_columns chỉ số dòng và số cột của dữ liệu

In [4]:
num_rows = len(ds_salary_df)
num_columns = len(ds_salary_df.columns)
num_rows, num_columns

(742, 41)

- Dữ liệu có 742 dòng và 42 cột

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

- Mỗi dòng có thông tin về mức lương tối thiểu, mức lương tối đa, mô tả về công việc, công ty,... liên quan đến nhà khoa học dữ liệu. Theo quan sát sơ bộ và mô tả về dữ liệu thì không có dòng nào có ý nghĩa khác.

### Dữ liệu có các dòng bị lặp không?

- Biến duplicated_rows dùng để kiểm tra có dòng lặp hay không bằng phương thức duplicated của pandas, trả về True nếu bị lặp, ngược lại là False.

In [5]:
duplicated_rows = ds_salary_df.duplicated().any()
duplicated_rows

True

- Do dữ liệu có dòng bị lặp nên mình tiến hành loại bỏ các dòng bị lặp, chỉ giữ lại dòng đầu tiên.

In [6]:
ds_salary_df = ds_salary_df.drop_duplicates()
ds_salary_df.shape

(467, 41)

### Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả về các cột trong file "data_scientist_salary":
- **Job title**: Chức danh của công việc, ví dụ: Data scientist, junior data scientist, senior data scientist, ...
- **Salary Estimate**: Khoảng lương và nguồn lấy thông tin.
- **Job Description**: Mô tả chung về công việc, cho biết các yêu cầu kỹ năng mà công ty mong muốn và những việc mà người trúng tuyển sẽ đảm nhiệm.
- **Rating**: Điểm đánh giá cho công ty.
- **Company Name**: Tên công ty.
- **Location**: Địa điểm làm việc.
- **Headquarters**: Địa điểm trụ sở chính của công ty.
- **Size**: Quy mô nhân lực của công ty.
- **Founded**: Năm thành lập công ty. 
- **Type of ownership**: Cho biết đây là công ty tư nhân, đại chúng hay thuộc nhà nước. 
- **Industry**: Cho biết hoạt động kinh doanh của công ty thuộc ngành công nghiệp nào, ví dụ: công nghệ thông tin, công nghiệp dược phẩm, ... 
- **Sector**: Lĩnh vực mà công ty hoạt động.
- **Revenue**: Tổng số doanh thu hàng năm của công ty. 
- **Competitors**: Cho biết những đối thủ cạnh tranh hiện tại của công ty. 
- **Hourly**: Có giá trị 1 nếu mức lương được báo cáo theo giờ và có giá trị là 0 ở các trường hợp còn lại.
- **Employer provided**: Có giá trị là 1 nếu mức lương được cung cấp bởi nhà tuyển dụng và có giá trị là 0 ở các trường hợp còn lại.
- **Lower Salary**: Mức lương tối thiểu cho công việc này của công ty.
- **Upper Salary**: Mức lương tối đa cho công việc này của công ty.
- **Avg Salary(K)**: Mức lương trung bình cho công việc này của công ty.
- **company_txt**: Chứa tên công ty.
- **Job Location**: Tên viết tắt của vị trí công ty (bang).
- **Age**: Tuổi của công ty (tính theo năm).
- **Python -> google_an**: Có giá trị là 1 nếu công việc đòi hỏi kỹ năng ..., ngược lại có giá trị 0.
- **job_title_sim**: Tên đơn giản của chức danh công việc.
- **seniority_by_title**: Có chữ senior trong chức danh.
- **Degree**: Có giá trị M nếu công việc yêu cầu bằng cấp hoặc số năm kinh nghiệm; có giá trị P trong các trường hợp còn lại.

**Nhận xét**
- Ở đây có quá nhiều cột dữ liệu, vì vậy để dễ dàng theo dõi và nhận xét thì ta sẽ bỏ bớt các cột không cần thiết ví dụ như:
    - Bỏ các cột có thông tin ngoài lề như: **Headquarters**(vị trí trụ sở), **Competitors**(Đối thủ cạnh tranh của cty), **Hourly**(lương theo giờ hay không), **Employer provided**(thông tin do nhân viên cung cấp hay không).
    - Bỏ cột trùng dữ liệu: **company_txt**(tên của công ty) trùng với Company Name.

In [7]:
ds_salary_df = ds_salary_df.drop(columns = ['Headquarters','Competitors','Hourly', \
                                            'Employer provided','company_txt'])
ds_salary_df.shape

(467, 36)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

- Dùng chức năng dtypes để xem kiểu dữ liệu của các cột.

In [8]:
col_dtypes = ds_salary_df.dtypes
col_dtypes

Job Title              object
Salary Estimate        object
Job Description        object
Rating                float64
Company Name           object
Location               object
Size                   object
Founded                 int64
Type of ownership      object
Industry               object
Sector                 object
Revenue                object
Lower Salary            int64
Upper Salary            int64
Avg Salary(K)         float64
Job Location           object
Age                     int64
Python                  int64
spark                   int64
aws                     int64
excel                   int64
sql                     int64
sas                     int64
keras                   int64
pytorch                 int64
scikit                  int64
tensor                  int64
hadoop                  int64
tableau                 int64
bi                      int64
flink                   int64
mongo                   int64
google_an               int64
job_title_

- Các cột có kiểu dữ liệu đã phù hợp ngoại trừ cột **Founded**: dùng để chỉ năm thành lập của công ty, nên chuyển thành datetime với format là '%Y'.
- Tuy nhiên khi xử lý sẽ gặp lỗi bởi vì các giá trị thiếu được điền là '-1'. Vì vậy để dễ xử lý ta cần thay đổi các giá trị '-1' thành np.NaN.

In [9]:
ds_salary_df = ds_salary_df.replace(-1, np.NaN)
ds_salary_df['Founded'] = pd.to_datetime(ds_salary_df['Founded'],format = '%Y')
ds_salary_df['Founded'].head()

0   1973-01-01
1   1984-01-01
2   2010-01-01
3   1965-01-01
4   1998-01-01
Name: Founded, dtype: datetime64[ns]

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?
(tính tỉ lệ các giá trị thiếu, min/max của các cột có kiểu dữ liệu dạng số)

Vì các cột từ Python -> google_an: chỉ gồm 1 nếu công việc yêu cầu skill này hoặc ngược lại là 0 cho nên ta không cần xét sự phân bố các cột này. Các cột có kiểu dữ liệu dạng số (numerical) cần xét còn lại là: **"Rating", "Lower Salary", "Upper Salary", "Avg Salary(K)", "Age"**. 

Với mỗi cột này ta tính phần trăm các giá trị thiếu, min, max bằng cách dùng cấu trúc agg để tổng hợp các tính năng tính giá trị trong mỗi cột.

In [10]:
numerical_df = ds_salary_df[["Rating", "Lower Salary", "Upper Salary", "Avg Salary(K)", "Age"]]
def missing_ratio(x):
    return x.isna().mean() * 100

numerical_df = numerical_df.agg([missing_ratio, pd.Series.min, pd.Series.max])
numerical_df

Rating  Lower Salary  Upper Salary  Avg Salary(K)         Age
missing_ratio  1.498929           0.0           0.0            0.0    7.066381
min            1.900000          15.0          16.0           15.5    2.000000
max            5.000000         202.0         306.0          254.0  277.000000

- Ta có thể thấy min/max các cột không có gì bất thường, ngoại trừ Age của 1 công ty có tuổi '277'?

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?
(tính tỉ lệ các giá trị thiếu, số lượng các giá trị khác nhau, show một vài giá trị)

Theo như quan sát và tìm hiểu dữ liệu thì các cột categorical của dữ liệu trên sẽ là các cột có dtypes là object.

Tuy nhiên: 
- Ở các cột **seniority_by_title**, **Degree** thì các giá trị thiếu được điền là na -> cần chuyển thành np.NaN để có thể hoạt động tốt trên các hàm của pandas.
- Ở cột **Revenue** thì các giá trị thiếu được điền là Unknown -> cần chuyển thành np.NaN

In [11]:
ds_salary_df = ds_salary_df.replace(['na', 'Unknown / Non-Applicable'], np.NaN)

In [12]:
columns_cate = list(ds_salary_df.dtypes[ds_salary_df.dtypes == 'object'].index)
categorical_df = ds_salary_df[columns_cate]
def diff_vals(s):
    return s.dropna().unique()
def num_diff_vals(s):
    return s.nunique()
categorical_df = categorical_df.agg([missing_ratio,num_diff_vals,diff_vals])
categorical_df

Job Title  \
missing_ratio                                                0.0   
num_diff_vals                                                264   
diff_vals      [Data Scientist, Healthcare Data Scientist, Re...   

                                                 Salary Estimate  \
missing_ratio                                                0.0   
num_diff_vals                                                416   
diff_vals      [$53K-$91K (Glassdoor est.), $63K-$112K (Glass...   

                                                 Job Description  \
missing_ratio                                                0.0   
num_diff_vals                                                463   
diff_vals      [Data Scientist\nLocation: Albuquerque, NM\nEd...   

                                                    Company Name  \
missing_ratio                                                0.0   
num_diff_vals                                                343   
diff_vals      [Tecolote Research\n3.8, University of Marylan...   

                                                        Location  \
missing_ratio                                                0.0   
num_diff_vals                                                200   
diff_vals      [Albuquerque, NM, Linthicum, MD, Clearwater, F...   

                                                            Size  \
missing_ratio                                                0.0   
num_diff_vals                                                  8   
diff_vals      [501 - 1000 , 10000+ , 1001 - 5000 , 51 - 200 ...   

                                               Type of ownership  \
missing_ratio                                                0.0   
num_diff_vals                                                  9   
diff_vals      [Company - Private, Other Organization, Govern...   

                                                        Industry  \
missing_ratio                                                0.0   
num_diff_vals                                                 60   
diff_vals      [Aerospace & Defense, Health Care Services & H...   

                                                          Sector  \
missing_ratio                                                0.0   
num_diff_vals                                                 25   
diff_vals      [Aerospace & Defense, Health Care, Business Se...   

                                                         Revenue  \
missing_ratio                                           28.69379   
num_diff_vals                                                 12   
diff_vals      [$50 to $100 million (USD), $2 to $5 billion (...   

                                                    Job Location  \
missing_ratio                                                0.0   
num_diff_vals                                                 37   
diff_vals      [NM, MD, FL, WA, NY, TX, CA, VA, MA, NJ, CO, I...   

                                                   job_title_sim  \
missing_ratio                                           1.284797   
num_diff_vals                                                  9   
diff_vals      [data scientist, other scientist, analyst, dat...   

              seniority_by_title     Degree  
missing_ratio          72.805139  52.462527  
num_diff_vals                  2          2  
diff_vals               [sr, jr]     [M, P]

## 3. Đưa ra các câu hỏi có ý nghĩa cần trả lời

### Câu hỏi 1: 

- Trả lời được có lợi ích gì?

### Câu hỏi 2:

- Trả lời được có lợi ích gì?

## 4. Tiền xử lý và phân tích dữ liệu để trả lời các câu hỏi

## Câu hỏi 1:

### Tiền xử lý
- Các bước thực hiện:

In [14]:
#Code

### Phân tích dữ liệu
- Các bước thực hiện:

In [15]:
#Code

## Câu hỏi 2:

### Tiền xử lý
- Các bước thực hiện:

In [16]:
#Code

### Phân tích dữ liệu
- Các bước thực hiện:

In [17]:
#Code

## 5. Nhìn lại quá trình làm đồ án

- TV1 : Những khó khăn?
        Học được gì?

- TV2: Những khó khăn?
        Học được gì?

- Nhóm: nếu có thêm thời gian thì sẽ làm gì?

## 6. Tài liệu tham khảo